core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -259,21 +259,29 @@ private[spark] class TaskSetManager(
       loc match {
         case e: ExecutorCacheTaskLocation =>
           pendingTaskSetToAddTo.forExecutor.getOrElseUpdate(e.executorId, new ArrayBuffer) += index
+          pendingTaskSetToAddTo.forHost.getOrElseUpdate(loc.host, new ArrayBuffer) += index
         case e: HDFSCacheTaskLocation =>
           val exe = sched.getExecutorsAliveOnHost(loc.host)
           exe match {
             case Some(set) =>
               for (e <- set) {
                 pendingTaskSetToAddTo.forExecutor.getOrElseUpdate(e, new ArrayBuffer) += index
               }
+              pendingTaskSetToAddTo.forHost.getOrElseUpdate(loc.host, new ArrayBuffer) += index
               logInfo(s"Pending task $index has a cached location at ${e.host} " +
                 ", where there are executors " + set.mkString(","))
             case None => logDebug(s"Pending task $index has a cached location at ${e.host} " +
               ", but there are no executors alive there.")
           }
-        case _ =>
+        case _: HostTaskLocation =>
+          val exe = sched.getExecutorsAliveOnHost(loc.host)
+          exe match {
+            case Some(_) =>
+              pendingTaskSetToAddTo.forHost.getOrElseUpdate(loc.host, new ArrayBuffer) += index
+            case _ => logDebug(s"Pending task $index has a location at ${loc.host} " +
+              ", but there are no executors alive there.")
+          }
       }
-      pendingTaskSetToAddTo.forHost.getOrElseUpdate(loc.host, new ArrayBuffer) += index
```

Comment:
Irrespective of whether we have an executor on a host or not, we should update.
An executor could be added later on to that host.

Reply:
@mridulm Thank you for your review.
I recognize the downside of doing this, causing many possibilities to fail that could be accelerated with a little waiting.

The trouble for me is that the host doesn't have a woker or NodeManager, so it can never start an executor on the corresponding host machine. Also the `getAllowedLocalityLevel` method prevents the task from trying to go farther locally. The task can never be actually executed.

Based on this, I have two options in mind
1. can we implement a new task locality, e.g. `HDFSDiskCacheTaskLocation` instead of `HostTaskLocation` here.
Then only the executor alive determination will be done for this case.
https://github.com/apache/spark/blob/ffe3fc9d23967e41092cf67539aa7f0d77b9eb75/core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala#L456-L460

2. or modify the implementation of `getAllowedLocalityLevel` to allow farther locality with limited attempts

Reply:
Two things:

a) The host might not have a worker right now - but that can change as the application executes, and either new executors are added (dynamic resource allocation) or if executors fail.

b) `computeValidLocalityLevels` will readjust the valid locality levels based on what scheduler knows about.

I am not sure what you mean by "task can never be actually executed".
Please take a look at `getAllowedLocalityLevel` on how locality level is computed.

A task will always be added to `all` tasks - and also to some rack typically: so it will always get executed, though might be slightly delayed if other tasks for the stage can have better schedule possible

Reply:
I'm sorry I may not have expressed the situation I encountered clearly
Please Take a look at this example: It query a hive partition.
`dataset.persist(StorageLevel.MEMORY_AND_DISK())` is an important statement that makes the task accompanied by a preferred locations.
```java
// The online environment is actually hive partition data imported to tidb, the code logic can be simplified as follows
SparkSession testApp = SparkSession.builder()
    .master("local[*]")
    .appName("test app")
    .enableHiveSupport()
    .getOrCreate();
Dataset<Row> dataset = testApp.sql("select * from default.test where dt = '20211129'");
dataset.persist(StorageLevel.MEMORY_AND_DISK());
dataset.count();
```

Here is the forHost of my hook runtime
```
forHost=@HashMap[
        serialVersionUID=@Long[1],
        _loadFactor=@Integer[750],
        table=@HashEntry[][
            @DefaultEntry[(kv: hdfs-loc-1, ArrayBuffer(2, 1))],
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            @DefaultEntry[(kv: driver-host, ArrayBuffer())],
            null,
            null,
            null,
            null,
            @DefaultEntry[(kv: hdfs-loc-2, ArrayBuffer(2, 1))],
            @DefaultEntry[(kv: hdfs-loc-3, ArrayBuffer(2, 1))],
        ],
```
Task 1, 2 preferred location is HostTaskLocation , pointing to `hdfs-loc-1,  hdfs-loc-2, hdfs-loc-3`
But these three hosts are hdfs data blocks where the machine, these three machines do not exist spark worker, there is no nodeManager. `getAllowedLocalityLevel` will only return `NODE_LOCAL` in this case. So "task can never be actually executed".

Reply:
Couple of things:

a) Tasks expressing preferred locality does not necessarily mean spark will wait at that preferred level. Take a look at `computeValidLocalityLevels`.
In your case, if scheduler is waiting for NODE_LOCAL schedules, then it means there is some executor which is alive on some host which is preferred.

b) If a task is not run at one level, it would at the next level - after a modest delay which is configurable.

Reply:
@mridulm, thanks for the detailed answer, I currently avoid task pending by setting spark.locality.wait.node to 0. In fact a permanent pending is possible, if there are not many resources, spark.locality.wait.node defaults to 3s, and all remaining resources are tried within this time range, then there is no chance to get to the next TaskLocality, although `computeValidLocalityLevels` returns [ PROCESS_LOCAL, NODE_LOCAL, ANY].

While setting spark.locality.wait.node to 0 eased my production environment's trouble, I think it would be better to treat the TaskLocation special in this case, as the current code also treats the HDFSCacheTaskLocation special
https://github.com/apache/spark/blob/030de1d09f121b167aaaa8237a2807f902c1e710/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L262-L272

Reply:
> b) If a task is not run at one level, it would at the next level - after a modest delay which is configurable.
It does not result in task never getting executed.

Add some spark ui screenshots, I reproduce the situation, The task is very light, 100,000 rows of data computing count
![image](https://user-images.githubusercontent.com/4069905/144710209-083fe6c4-b6af-40a8-a65c-b360505f5b9e.png)
![image](https://user-images.githubusercontent.com/4069905/144710235-d9824a94-ed3e-47de-a0bf-41550ea52d3d.png)
![image](https://user-images.githubusercontent.com/4069905/144710495-0c81e7f8-3405-4c9a-bd67-a3c9db3a2a3e.png)
![image](https://user-images.githubusercontent.com/4069905/144710249-7e7226c1-44fa-411a-8914-87c874cf4c24.png)

Reply:
> In fact a permanent pending is possible, if there are not many resources, spark.locality.wait.node defaults to 3s, and all remaining resources are tried within this time range, then there is no chance to get to the next TaskLocality, although computeValidLocalityLevels returns [ PROCESS_LOCAL, NODE_LOCAL, ANY].

I am confused about this statement - if there are no resources available, no tasks can be scheduled and so spark will  wait indefinitely.

If/when there are resources available, scheduler will move to the next locality level once the previous level times out.

> I think it would be better to treat the TaskLocation special in this case, as the current code also treats the HDFSCacheTaskLocation special

Current behavior of `HDFSCacheTaskLocation` is:
a) If there are executors on the host, add to `forExecutor`
b) Add to `forHost`

(b) is done for all tasks which have locality preference.

As and when a new executor is added or an existing executor is removed, `computeValidLocalityLevels` iterates through the various `for<LEVEL>` to identify if there are executors available on that LEVEL to determine whether to keep the level for schedule.

What is proposed in the PR will break `computeValidLocalityLevels` - since addition of a new executor will not add host level schedule.

If the issue we are attempting to solve is:
"for the current pending tasks, with the executor resources available to the application, disregard a locality level if there are no valid candidates for that level"

If yes, the current PR is not solving this issue - though this is a good problem to address in the scheduler.
We will have to look at the cost/benefit of doing that based on the proposal which fixes this issue - since I suspect the costs would not be very cheap (the various materialized datastructures in scheduler are to keep the costs as low as possible).

Hope that clarifies.

Reply:
Yes, I totally agree with you and after discussing with you I realized that this pr just solves my special case and breaks other cases.

The reason I haven't closed this pr yet is because I still want to find a way to fix this situation with minimal modifications, while not affecting other scenarios that are already working well.

@mridulm, thank you for spending a lot of time discussing this and giving me new insight into spark scheduling as well.